File: benchmarks/bst/java/Bst.java
import java.util.Random;

public class Bst {
    static class Node {
        long val;
        Node left, right;
        Node(long val) { this.val = val; }
    }

    static void insert(Node root, long val) {
        Node cur = root;
        while (true) {
            if (val < cur.val) {
                if (cur.left == null) { cur.left = new Node(val); return; }
                cur = cur.left;
            } else {
                if (cur.right == null) { cur.right = new Node(val); return; }
                cur = cur.right;
            }
        }
    }

    static long prev;
    static boolean first = true;
    static boolean ok = true;
    static long maxVal;
    static long count = 0;

    static void inorder(Node node) {
        if (node == null) return;
        inorder(node.left);
        if (!first && node.val < prev) ok = false;
        first = false;
        prev = node.val;
        maxVal = node.val;
        count++;
        inorder(node.right);
    }

    public static void main(String[] args) {
        long n = args.length > 0 ? Long.parseLong(args[0]) : 300_000;

        Random rnd = new Random(42);
        Node root = new Node(rnd.nextLong() & Long.MAX_VALUE);
        for (long i = 1; i < n; i++) {
            insert(root, rnd.nextLong() & Long.MAX_VALUE);
        }

        inorder(root);

        if (!ok || count != n) {
            System.err.println("self-check failed: in-order traversal not sorted or count mismatch");
            System.exit(1);
        }

        System.out.println(maxVal);
    }
}

File: benchmarks/bst/c/bst.c
#include <stdio.h>
#include <stdlib.h>

static unsigned int lcg_state;

static long lcg_next(void) {
    lcg_state = (lcg_state * 1103515245u + 12345u) & 0x7fffffffu;
    return (long)lcg_state;
}

typedef struct Node {
    long val;
    struct Node *left, *right;
} Node;

Node *make_node(long val) {
    Node *n = malloc(sizeof(Node));
    n->val = val;
    n->left = n->right = NULL;
    return n;
}

void insert(Node **root, long val) {
    Node **cur = root;
    while (*cur != NULL) {
        if (val < (*cur)->val) cur = &(*cur)->left;
        else cur = &(*cur)->right;
    }
    *cur = make_node(val);
}

void inorder(Node *node, long *prev, int *first, int *ok, long *max_val, long *count) {
    if (node == NULL) return;
    inorder(node->left, prev, first, ok, max_val, count);
    if (!*first && node->val < *prev) *ok = 0;
    *first = 0;
    *prev = node->val;
    *max_val = node->val;
    (*count)++;
    inorder(node->right, prev, first, ok, max_val, count);
}

int main(int argc, char **argv) {
    long n = argc > 1 ? atol(argv[1]) : 300000;

    lcg_state = 42;
    Node *root = NULL;
    for (long i = 0; i < n; i++) {
        insert(&root, lcg_next());
    }

    long prev = 0, max_val = 0;
    int first = 1, ok = 1;
    long count = 0;
    inorder(root, &prev, &first, &ok, &max_val, &count);

    if (!ok || count != n) {
        fprintf(stderr, "self-check failed: in-order traversal not sorted or count mismatch\n");
        return 1;
    }

    printf("%ld\n", max_val);
    return 0;
}
